components/store_state.py

```diff
@@ -26,47 +28,169 @@ def load_state_from_store(store_data, key):
+                        if model in ["single", "range", "compare"]:
+                            store_data = StoreState.save_state_to_store(store_data, "model", model)
+                        else:
+                            store_data = StoreState.save_state_to_store(store_data, "model", None)
+                        store_data = StoreState.save_state_to_store(store_data, "type", None)
+
+                    for key, value in query_params.items():
+                        value = value[0]
+                        store_data = StoreState.save_state_to_store(store_data, key, value)
+
+            return store_data
 
         @app.callback(
-            Output({"type": "dynamic-input", "index": ALL}, "value"),
+            [Output(id, "value") for id in self.persist_ids],
             Input(store_id, "modified_timestamp"),
             State(store_id, "data"),
             prevent_initial_call=True,
         )
         def load_data(modified_timestamp, store_data):
             if modified_timestamp:
+
+                if store_data is None or not isinstance(store_data, dict):
+                    store_data = {}
+
                 # Get the current number of dynamic input components
                 num_outputs = len(ctx.outputs_list)
 
-                # Initialize output list, default to using no_update
+                # Initialize the output list, defaulting to no_update
                 input_values = [no_update] * num_outputs
 
-                # Update existing data
+                # If stored data exists, update the corresponding component values based on the indices
                 if store_data:
-                    for i in range(num_outputs):
-                        input_values[i] = store_data.get(f"input_{i}", no_update)
+                    # Iterate through the key-value pairs in the stored data
+                    for key, value in store_data.items():
+                        # Iterate through all the output components, matching the stored data indices
+                        for i, output in enumerate(ctx.outputs_list):
+                            # If the component's id matches the key in the stored data, update the component's value
+                            if output["id"] == key:
+                                input_values[i] = value
 
                 return input_values
 
-            # If modified_timestamp is empty or invalid, return a list of no_update
+            # If modified_timestamp is empty or invalid, return a list with no_update
             return [no_update] * len(ctx.outputs_list)
+
+        # New feature, reflect to the URL
+        @app.callback(
+            Output("url", "href"),
+            [Input(id, "value") for id in self.persist_ids],
+            [Input("url", "href")],
+            State(store_id, "data"),
+            prevent_initial_call=True,
+        )
+        def update_url(*args):
+            # The URL is the second-to-last argument
+            url = args[-2]
```

Comment:
this is a code duplication, consider creating a function

Reply:
```
def update_url(*args):
    *id, url, store_data = args
```
Now the code has been changed to this